Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp

```diff
@@ -65,71 +52,42 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
         paintGraphicsLayerContents(context, mappedTileRect);
     };
 
-    auto paintBuffer = [&](Nicosia::Buffer& buffer) {
-        buffer.beginPainting();
+    // Threaded rendering: record display lists, and asynchronously replay them using dedicated worker threads.
+    WTFBeginSignpost(this, RecordTile);
 
+    auto displayList = makeUnique<DisplayList::DisplayList>();
+    DisplayList::RecorderImpl recordingContext(*displayList, GraphicsContextState(), FloatRect({ }, tileRect.size()), AffineTransform());
+    paintIntoGraphicsContext(recordingContext);
+
+    auto* threadedPaintingPool = m_coordinator->skiaThreadedPaintingPool();
+    ASSERT(threadedPaintingPool);
+
+    auto flags = Nicosia::Buffer::Flags { contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha };
+    auto buffer = threadedPaintingPool->postPaintingTask(tileRect.size(), flags, [displayList = WTFMove(displayList), tileRect](Nicosia::Buffer& buffer) mutable {
         RELEASE_ASSERT(buffer.surface());
         if (auto* canvas = buffer.surface()->getCanvas()) {
```

Comment:
If the buffer has a surface it should have a canvas too.

Reply:
Actually, I find beginThreadedGPUPainting a bit confusing, it's only needed for GPU and it doesn't have an endThreadedGPUPainting. What if we rename it as `canvas()` and make it pure virtual so that in case of non-accelerated buffer it just returns the can canvas and in the accelerated one it creates the buffer and returns the canvas? In that case we still need to check here that it's not nullptr.

Reply:
Nice idea, fixed.